h8mir-BJS2-87681

Deathroels opened the pull request.

Deathroels committed:
start working on BJS2-87681

Deathroels committed:
BJS2-87681:
-Добавил регистрацию обработчика оповещений
-Необходимо найти ошибку в компиляции и продолжить задачу на 4-м пункте

Deathroels committed:
Merge remote-tracking branch 'origin/h8mir-BJS2-87681' into h8mir-BJS2-87681

# Conflicts:
#	src/main/java/school/faang/bjs2_87681/Main.java
#	src/main/java/school/faang/bjs2_87681/Notification.java
#	src/main/java/school/faang/bjs2_87681/NotificationManager.java

Deathroels committed:
BJS2-87681:
-Добавил регистрацию обработчика оповещений
-Необходимо найти ошибку в компиляции и продолжить задачу на 4-м пункте

Deathroels committed:
BJS2-87681:
-Добавил регистрацию обработчика оповещений
-была проблема с билдом, продолжить на 4-ом пунтке задачи

Deathroels committed:
BJS2-87681:
-Добавлена фильтрация сообщений перед отправкой

Deathroels committed:
Merge remote-tracking branch 'origin/h8mir-BJS2-87681' into h8mir-BJS2-87681

# Conflicts:
#	src/main/java/school/faang/bjs2_87681/Notification.java

iBerbasov commented on src/main/java/school/faang/bjs2_87681/Main.java:
```diff
@@ -0,0 +1,30 @@
+public class Main {
+    public static void main(String[] args) {
+        NotificationManager notificationManager = new NotificationManager();
+        // Регистрация обработчиков оповещений
```
Комменты в коде - зло. они всегда приносят больше вреда, чем пользы
Понятно, что тут они перетекают из примера в условии задачи, но свой финальный код всегда стоит приводить в чистый вид без всякого лишнего

iBerbasov commented on src/main/java/school/faang/bjs2_87681/NotificationManager.java:
```diff
@@ -0,0 +1,42 @@
+
+@ToString
+@Getter
+@EqualsAndHashCode
```
А зачем менеджеру все эти аннотации?
Не стоит добавлять лишнего там, где это не требуется

iBerbasov commented on src/main/java/school/faang/bjs2_87681/NotificationManager.java:
```diff
@@ -0,0 +1,42 @@
+        if (notification == null) {
+            throw new IllegalArgumentException("null передан как объект Notification");
+        }
+        if (notification.getMessage() == null) { // в итоге кажется что эта проверка и не нужна
```
В целом почему нет - отправлять пустые сообщения не всегда хочется
Если говорить про npe, то да, не вижу в коде проблем, если сообщение будет нуллом

iBerbasov commented on src/main/java/school/faang/bjs2_87681/NotificationManager.java:
```diff
@@ -0,0 +1,42 @@
+@Getter
+@EqualsAndHashCode
+public class NotificationManager {
+    private final Map<NotificationType, Consumer<Notification>> notificationTypeConsumerMap = new HashMap<>();
```
Не очень информативное имя переменной, лучше давать имена ближе к объектной области - здесь хранятся обработчики сообщений, вот и назвать её лучше чем-то вроде  handlers, notificationHandlers  или handlersMap

iBerbasov commented on src/main/java/school/faang/bjs2_87681/NotificationManager.java:
```diff
@@ -0,0 +1,42 @@
+        notificationTypeConsumerMap.get(notification.getType()).accept(notification);
+    }
+
+    public void isContainsBadWords(Notification notification, Predicate<String> filter) {
```
Возможно у тебя и был такой замысел, но выглядит не слишком очевидно
В чем смысл этой проверки и фильтра плохих слов? 
Если в сообщении попадается оскорбление, то выполнение кода не прерывается и сообщение все равно отправляется 

Мы вас оскорбим в любом случае, но если что, то в логах предупредим об этом ))
В целом тоже имеет место быть и такое поведение ))

Deathroels committed:
BJS2-87681: fix
-переместил валидацию полей класса, в сам класс Notification
-исправил логику работы метода validateBadWords, в соответствии с комментарием ревью
-оставил рефактор кода на следующий ПР с метками todo

Deathroels commented:
В одной из задач получил комментарий на будущее, что "тех долги, отправлять отдельным ПРом", правильно ли я понял, 
это значит фиксы, баги, нарушения логики и т.п. корректно отправлять в этом же ПРе, для того чтобы не затягивать работающий код в работе, а уже правки и изменения работающего кода можно отправить другим ПРом?
К примеру:
Первоначальный ПР
Комментарий по пропущенному багу или функциональности
новый коммит и пуш (который автоматически прикрепится к первоначальному ПРу)
мерж (предположим со стороны лида) ветки с фиксом в выбранную ветку, куда я потом опять же из своей ветки по данной задаче, отправляю новый ПР с незначительными для работы фиксами, но требующими доработки?

Deathroels committed:
BJS2-87681: refactor
-review comments

iBerbasov commented:
По решению ок, по вопросу:
В целом да, особенно в случае наших небольших задачек - тут вообще не применимо понятие тех.долга, т.к. задачи маленькие и независимые друг от друга

Ситуация может отличаться в разных задачах. Пропущенный баг? Баг может быть критичным или некритичным. Если баг некритичный и код с этим багом можно влить в основную ветку, то да - образуется тех.долг и его можно будет решать когда-то позже (_ха-ха, обычно никогда_) в других ПР-ах. Баг критичный - естественно такой код вливать нельзя, он сломает всё. Это уже надо будет решать в текущем ПР

Это будет актуально при работе над общим большим проектом. Ну а как в деталях этот процесс выстраивать - вопрос извечный и в каждой команде\компании\организации над этим всегда ломают голову ))
Правильного ответа на него нет и не будет - у всех свои нюансы работы, под которые надо подстраивать свои процессы. Если интересно можешь почитать на досуге о git flow, github flow, gitlab flow и еще куче их последователей

Как процесс выстроите вы в своей команде - не знаю, но думаю вы быстро скорректируетесь уже непосредственно с ментором и командой
